integration/proxy_tunnel_strategy_test.go

```diff
@@ -0,0 +1,418 @@
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/google/uuid"
+	"github.com/jackc/pgconn"
+	"github.com/stretchr/testify/require"
+)
+
+type proxyTunnelStrategy struct {
+	username string
+	cluster  string
+
+	lb      *utils.LoadBalancer
+	auth    *TeleInstance
+	proxies []*TeleInstance
+	node    *TeleInstance
+
+	db           *TeleInstance
+	dbAuthClient *auth.Client
+	postgresDB   *postgres.TestServer
+}
+
+// TestProxyTunnelStrategyAgentMesh tests the agent-mesh tunnel strategy
+func TestProxyTunnelStrategyAgentMesh(t *testing.T) {
+	lib.SetInsecureDevMode(true)
+	defer lib.SetInsecureDevMode(false)
+
+	p := &proxyTunnelStrategy{
+		cluster:  "proxy-tunnel-agent-mesh",
+		username: mustGetCurrentUser(t).Username,
+	}
+
+	strategy := &types.TunnelStrategyV1{
+		Strategy: &types.TunnelStrategyV1_AgentMesh{
+			AgentMesh: types.DefaultAgentMeshTunnelStrategy(),
+		},
+	}
+
+	// boostrap a load balancer for proxies.
+	p.makeLoadBalancer(t)
+
+	// boostrap an auth instance.
+	p.makeAuth(t, strategy)
+
+	// boostrap two proxy instances.
+	p.makeProxy(t)
+	p.makeProxy(t)
+	require.Len(t, p.proxies, 2)
+
+	// boostrap a node instance.
+	p.makeNode(t)
+
+	// boostrap a db instance.
+	p.makeDatabase(t)
+
+	// wait for the node and database to open reverse tunnels to both proxies.
+	waitForActiveTunnelConnections(t, p.proxies[0].Tunnel, p.cluster, 2)
+	waitForActiveTunnelConnections(t, p.proxies[1].Tunnel, p.cluster, 2)
+
+	// make sure we can connect to the node going though any proxy.
+	p.dialNode(t)
+
+	// make sure we can connect to the database going though any proxy.
+	p.dialDatabase(t)
+}
+
+// TestProxyTunnelStrategyProxyPeering tests the proxy-peer tunnel strategy
+func TestProxyTunnelStrategyProxyPeering(t *testing.T) {
+	lib.SetInsecureDevMode(true)
+	defer lib.SetInsecureDevMode(false)
+
+	p := &proxyTunnelStrategy{
+		cluster:  "proxy-tunnel-proxy-peer",
+		username: mustGetCurrentUser(t).Username,
+	}
+
+	strategy := &types.TunnelStrategyV1{
+		Strategy: &types.TunnelStrategyV1_ProxyPeering{
+			ProxyPeering: types.DefaultProxyPeeringTunnelStrategy(),
+		},
+	}
+
+	// boostrap a load balancer for proxies.
+	p.makeLoadBalancer(t)
+
+	// boostrap an auth instance.
+	p.makeAuth(t, strategy)
+
+	// boostrap the first proxy instance.
+	p.makeProxy(t)
+	require.Len(t, p.proxies, 1)
+
+	// boostrap a node instance.
+	p.makeNode(t)
+
+	// boostrap a db instance.
+	p.makeDatabase(t)
+
+	// wait for the node to open a reverse tunnel to the first proxy.
+	waitForActiveTunnelConnections(t, p.proxies[0].Tunnel, p.cluster, 2)
+
+	// boostrap the second proxy instance after the node has already established
+	// a reverse tunnel to the first proxy.
+	p.makeProxy(t)
+	require.Len(t, p.proxies, 2)
+
+	// make sure node doesn't open any reverse tunnel to the second proxy.
+	waitForMaxActiveTunnelConnections(t, p.proxies[1].Tunnel, p.cluster, 0)
+
+	// make sure we can connect to the node going though any proxy.
+	p.dialNode(t)
```

Comment:
We discussed this the other day but figured we could discuss it more here. 

Does the peer to peer grpc connection cause the test to be flaky? Meaning if the client connection and dial request go through before the grpc connection is setup could this test fail? 

From what I remember we saw the test passing reliably. I think we should just double check that it is indeed reliable. I think the thing to validate is what happens if we use a grpc clientConn that is not established yet. Either it fails immediately in which case this could be flaky or it waits with some timeout for the connection to be established in which case we shouldn't need to do anything else.